### PR TITLE
Bump webpack 5

### LINKED ISSUE #4946: [heft] Rush stack heft plugins preventing webpack security upgrades
## Summary
Currently there is a webpack security issue with a released fix at 5.95.0.  However, the latest released versions of the heft plugins currently, through restrictive peer dependencies, do not cleanly allow users to upgrade past 5.82.

## Repro steps
From a rush stack monorepo, try to upgrade just webpack.  That'll force the usual cascading series of work where you also have to also upgrade a series of other rush stack pieces due to restrictive versioning (for better or worse).  In the end you get a bit blocked though because none of the rush stack heft plugins accommodate trying out newer tech.  Produces error like:
```
└─┬ @rushstack/heft-webpack5-plugin 0.10.12
  └── ✕ unmet peer webpack@~5.82.1: found 5.95.0
```

## Details
There are likely workarounds ranging from bad (setting up version overrides) to worse (removing strict peer dependencies altogether).
Perhaps the most realistic solution ATM would be to test and release plugin versions which allow the latest webpack.

## Standard questions

| Question | Answer |
| -------- | -------- |
| `@rushstack/heft` version? | 0.64.5 up to 0.67.2 |
| Operating system? | Windows and Mac |
| Would you consider contributing a PR? | No |
| Node.js version (`node -v`)? | v18.17.1 |

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
Fixes #4946 
Upgrades webpack 5 to 5.95.0.

## Details
Fixed various issues that cropped up as a result of the upgrade.

## How it was tested
All unit tests in the repository.
This PR's build policies.

## Impacted documentation
None

### CODE REVIEW

File: common/changes/@rushstack/heft-dev-cert-plugin/bump-webpack-5_2024-10-01-01-14.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Update to webpack 5.95.0",
```
Comment (dmichon-msft) [inline common/changes/@rushstack/heft-dev-cert-plugin/bump-webpack-5_2024-10-01-01-14.json:4]:
```suggestion
      "comment": "Remove unnecessary webpack dependency.",
```

File: webpack/webpack-workspace-resolve-plugin/src/test/createResolveForTests.ts [hunk 2]
```diff
@@ -16,7 +19,8 @@ export type WrappedResolve = (
   // eslint-disable-next-line @rushstack/no-new-null
 ) => [Error | false | null | undefined, ResolveRequest | undefined];
 
-export const parsedJson: Record<string, object> = {
+export type JsonObjectTypes = ReturnType<NonNullable<InputFileSystem['readJsonSync']>>;
```
Comment (D4N14L) [inline webpack/webpack-workspace-resolve-plugin/src/test/createResolveForTests.ts:22]: Import from the other file that exports it?

**Review by D4N14L — APPROVED** [review]

File: common/changes/@rushstack/heft-webpack5-plugin/bump-webpack-5_2024-10-01-01-14.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Update to webpack 5.95.0",
+      "type": "patch",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-webpack5-plugin/bump-webpack-5_2024-10-01-01-14.json:5]:
```suggestion
      "type": "minor",
```

File: common/changes/@rushstack/heft-webpack5-plugin/bump-webpack-5_2024-10-01-01-14.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Update to webpack 5.95.0",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-webpack5-plugin/bump-webpack-5_2024-10-01-01-14.json:4]:
```suggestion
      "comment": "Update the `webpack` peer dependency to `^5.82.1` from `~5.82.1`. Also bump `webpack-dev-server` to `^5.1.0`. This drops support for Node 16 and includes some breaking configuration changes. See https://github.com/webpack/webpack-dev-server/blob/master/migration-v5.md.",
```

File: common/changes/@rushstack/package-extractor/bump-webpack-5_2024-10-01-01-14.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Update to webpack 5.95.0",
+      "type": "patch",
```
Comment (iclanton) [inline common/changes/@rushstack/package-extractor/bump-webpack-5_2024-10-01-01-14.json:5]:
```suggestion
      "comment": "",
      "type": "none",
```

File: heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts [hunk 6]
```diff
@@ -334,15 +334,18 @@ export default class Webpack5Plugin implements IHeftTaskPlugin<IWebpackPluginOpt
...
-        compiler.hooks.infrastructureLog.tap(PLUGIN_NAME, (name: string, type: string, args: any[]) => {
-          if (name === WEBPACK_DEV_MIDDLEWARE_PACKAGE_NAME && type === 'error') {
-            const error: Error | undefined = args[0];
-            if (error) {
-              taskSession.logger.emitError(error);
+        compiler.hooks.infrastructureLog.tap(
+          PLUGIN_NAME,
+          (name: string, type: string, args: unknown[] | undefined) => {
+            if (name === WEBPACK_DEV_MIDDLEWARE_PACKAGE_NAME && type === 'error') {
+              const error: Error | undefined = args?.[0] as Error;
```
Comment (iclanton) [inline heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts:341]:
```suggestion
              const error: Error | undefined = args?.[0] as Error | undefined;
```

File: libraries/rush-lib/webpack.config.js [hunk 7]
```diff
@@ -82,7 +82,7 @@ module.exports = () => {
       },
       [
         new DeepImportsPlugin({
-          path: `${__dirname}/temp/rush-lib-manifest.json`,
+          path: `${__dirname}/temp/build/webpack-dll/[name].json`,
```
Comment (iclanton) [inline libraries/rush-lib/webpack.config.js:85]: This change is surprising.

File: webpack/set-webpack-public-path-plugin/src/SetPublicPathPlugin.ts [hunk 8]
```diff
@@ -137,16 +137,21 @@ export class SetPublicPathPlugin extends SetPublicPathPluginBase {
...
+        }
+
         let regexName: string;
         if (regexpName) {
           regexName = regexpName;
           if (regexpIsTokenized) {
             regexName = regexName
-              .replace(/\[name\]/g, Text.escapeRegExp(this.chunk.name))
-              .replace(/\[hash\]/g, this.chunk.renderedHash || '');
+              .replace(/\[name\]/g, Text.escapeRegExp(`${chunk.name}`))
```
Comment (iclanton) [inline webpack/set-webpack-public-path-plugin/src/SetPublicPathPlugin.ts:150]:
```suggestion
              .replace(/\[name\]/g, Text.escapeRegExp(String(chunk.name)))
```

File: webpack/webpack-plugin-utilities/package.json [hunk 9]
```diff
@@ -34,6 +34,6 @@
     "@rushstack/heft": "workspace:*",
     "local-node-rig": "workspace:*",
     "@types/tapable": "1.0.6",
-    "webpack": "~5.82.1"
+    "webpack": "~5.95.0"
```
Comment (iclanton) [inline webpack/webpack-plugin-utilities/package.json:37]: Update to `^5.95.0` and call this out in the changelog? Should be a minor bump too.

File: webpack/webpack5-load-themed-styles-loader/src/test/testData/getCompiler.ts [hunk 10]
```diff
@@ -3,9 +3,9 @@
 
 import path from 'path';
 import webpack from 'webpack';
-import type { Compiler, Stats } from 'webpack';
-import { createFsFromVolume, Volume } from 'memfs';
-import type { ILoadThemedStylesLoaderOptions } from '../..';
+import type { Compiler, OutputFileSystem, Stats } from 'webpack';
+import { Volume } from 'memfs';
+import type { ILoadThemedStylesLoaderOptions } from '../../index.js';
```
Comment (iclanton) [inline webpack/webpack5-load-themed-styles-loader/src/test/testData/getCompiler.ts:8]:
```suggestion
import type { ILoadThemedStylesLoaderOptions } from '../../index';
```

**Review by iclanton — APPROVED** [review]

File: .gitignore [hunk 11]
```diff
@@ -36,6 +36,7 @@ build/Release/
 
 # Dependency directories
 node_modules/
+**/.storybook/node_modules
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply .gitignore:39]: At least on my codespace, git is tracking the `.storybook/node_modules` folders in various projects that use it.

File: webpack/set-webpack-public-path-plugin/src/test/testBase.ts [hunk 12]
```diff
@@ -43,8 +43,8 @@ export function testForPlugin(pluginName: string, getPlugin: () => webpack.Webpa
       plugins: [getPlugin()]
     });
 
-    compiler.inputFileSystem = memoryFileSystem;
-    compiler.outputFileSystem = memoryFileSystem;
+    compiler.inputFileSystem = memoryFileSystem as unknown as InputFileSystem;
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply webpack/set-webpack-public-path-plugin/src/test/testBase.ts:46]: Breaking change to the typings of `InputFileSystem` and `OutputFileSystem` in webpack that made the `Volume` object no longer satisfy the contract.

**Review by iclanton — APPROVED** [review]

**Pull request merged**